src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -546,4 +615,188 @@ private CompletableFuture<List<V>> setToValueCache(List<V> assembledValues, List
     private static <T> DispatchResult<T> emptyDispatchResult() {
         return (DispatchResult<T>) EMPTY_DISPATCH_RESULT;
     }
+
+    private class BatchObserverImpl implements BatchObserver<V> {
+        private final CompletableFuture<List<V>> valuesFuture;
+        private final List<K> keys;
+        private final List<Object> callContexts;
+        private final List<CompletableFuture<V>> queuedFutures;
+
+        private final List<K> clearCacheKeys = new ArrayList<>();
+        private final List<V> completedValues = new ArrayList<>();
+        private int idx = 0;
+        private boolean onErrorCalled = false;
+        private boolean onCompletedCalled = false;
+
+        private BatchObserverImpl(
+            CompletableFuture<List<V>> valuesFuture,
+            List<K> keys,
+            List<Object> callContexts,
+            List<CompletableFuture<V>> queuedFutures
+        ) {
+            this.valuesFuture = valuesFuture;
+            this.keys = keys;
+            this.callContexts = callContexts;
+            this.queuedFutures = queuedFutures;
+        }
+
+        @Override
+        public void onNext(V value) {
+            assert !onErrorCalled && !onCompletedCalled;
+
+            K key = keys.get(idx);
+            Object callContext = callContexts.get(idx);
+            CompletableFuture<V> future = queuedFutures.get(idx);
+            if (value instanceof Throwable) {
```

Comment:
I am not sure this works in reactive land.   The contract that you get a Throwable via `onNext` is not something to cover and I dont think this could should either

We should expect a value that can be cast to type `V` - or do we do this in other BartchLoader code ;)